File: backend/apps/export/models/planning.py
from django.db import models
from apps.core.db_utils import schema_table, cyrillic_collation


class WeeklyTruckAllocation(models.Model):
    """Daily truck count decision per destination category.

    One row per (season, week_number, year, day_of_week). Captures total
    planned kg and the split between Russia/Kazakhstan/Gapy Satys trucks.
    total_trucks_calc is stored (planned_kg / 18500) for fast reads.

    DDL: export.weekly_truck_allocations
    UNIQUE (season_id, week_number, year, day_of_week)
    """

    # === Identity ===
    season = models.ForeignKey('core.Season', on_delete=models.PROTECT, db_column='season_id')
    week_number = models.PositiveSmallIntegerField()
    year = models.PositiveSmallIntegerField()
    # 1=Monday through 6=Saturday
    day_of_week = models.PositiveSmallIntegerField()

    # === Planned volume ===
    total_planned_kg = models.DecimalField(max_digits=12, decimal_places=2, null=True, blank=True)
    # Stored computed value: total_planned_kg / 18500
    total_trucks_calc = models.DecimalField(max_digits=8, decimal_places=2, null=True, blank=True)

    # === Truck split ===
    russia_trucks = models.IntegerField(default=0)
    kazakhstan_trucks = models.IntegerField(default=0)
    gapy_satys_trucks = models.IntegerField(default=0)

    # === Audit ===
    decided_by = models.ForeignKey(
        'core.User', on_delete=models.SET_NULL, null=True, blank=True,
        db_column='decided_by', related_name='truck_allocations_decided',
    )
    created_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        db_table = schema_table('export', 'weekly_truck_allocations')
        unique_together = [('season', 'week_number', 'year', 'day_of_week')]
        ordering = ['year', 'week_number', 'day_of_week']
        constraints = [
            models.CheckConstraint(check=models.Q(russia_trucks__gte=0), name='chk_trk_russia_gte0'),
            models.CheckConstraint(check=models.Q(kazakhstan_trucks__gte=0), name='chk_trk_kz_gte0'),
            models.CheckConstraint(check=models.Q(gapy_satys_trucks__gte=0), name='chk_trk_gapy_gte0'),
        ]

    def __str__(self) -> str:
        return f'W{self.week_number}/{self.year} day={self.day_of_week} — {self.total_trucks_calc} trucks'


PLAN_STATUS_CHOICES = [
    ('draft', 'Draft'),
    ('submitted', 'Submitted'),
    ('approved', 'Approved'),
    ('rejected', 'Rejected'),
]

# Allowed transitions: from_status → list of to_status values.
PLAN_TRANSITIONS = {
    'draft': ['submitted'],
    'submitted': ['approved', 'rejected'],
    'rejected': ['submitted'],
    'approved': [],
}


class WeeklyHarvestPlan(models.Model):
    """AD-3: Weekly harvest plan per greenhouse block.

    One row per (season, block, week_number, year). Plan vs actual for Mon–Sat.
    Includes approval workflow: draft → submitted → approved / rejected.

    DDL: export.weekly_harvest_plans — UNIQUE (season_id, block_id, week_number, year)
    """

    # === Identity ===
    season = models.ForeignKey('core.Season', on_delete=models.PROTECT, db_column='season_id')
    block = models.ForeignKey('core.GreenhouseBlock', on_delete=models.PROTECT, db_column='block_id')
    week_number = models.PositiveSmallIntegerField()  # ISO week 1-53
    year = models.PositiveSmallIntegerField()

    # === Plan (kg) ===
    monday_plan_kg = models.DecimalField(max_digits=10, decimal_places=2, default=0)
    tuesday_plan_kg = models.DecimalField(max_digits=10, decimal_places=2, default=0)
    wednesday_plan_kg = models.DecimalField(max_digits=10, decimal_places=2, default=0)
    thursday_plan_kg = models.DecimalField(max_digits=10, decimal_places=2, default=0)
    friday_plan_kg = models.DecimalField(max_digits=10, decimal_places=2, default=0)
    saturday_plan_kg = models.DecimalField(max_digits=10, decimal_places=2, default=0)

    # === Actual (kg) ===
    monday_actual_kg = models.DecimalField(max_digits=10, decimal_places=2, null=True, blank=True)
    tuesday_actual_kg = models.DecimalField(max_digits=10, decimal_places=2, null=True, blank=True)
    wednesday_actual_kg = models.DecimalField(max_digits=10, decimal_places=2, null=True, blank=True)
    thursday_actual_kg = models.DecimalField(max_digits=10, decimal_places=2, null=True, blank=True)
    friday_actual_kg = models.DecimalField(max_digits=10, decimal_places=2, null=True, blank=True)
    saturday_actual_kg = models.DecimalField(max_digits=10, decimal_places=2, null=True, blank=True)

    # === Approval workflow ===
    status = models.CharField(max_length=20, choices=PLAN_STATUS_CHOICES, default='draft')
    submitted_at = models.DateTimeField(null=True, blank=True)
    submitted_by = models.ForeignKey(
        'core.User', on_delete=models.SET_NULL, null=True, blank=True,
        db_column='submitted_by', related_name='harvest_plans_submitted',
    )
    approved_at = models.DateTimeField(null=True, blank=True)
    approved_by = models.ForeignKey(
        'core.User', on_delete=models.SET_NULL, null=True, blank=True,
        db_column='approved_by', related_name='harvest_plans_approved',
    )
    rejected_at = models.DateTimeField(null=True, blank=True)
    rejected_by = models.ForeignKey(
        'core.User', on_delete=models.SET_NULL, null=True, blank=True,
        db_column='rejected_by', related_name='harvest_plans_rejected',
    )
    rejection_note = models.CharField(max_length=500, blank=True, null=True)

    # === Audit ===
    entered_by = models.ForeignKey(
        'core.User', on_delete=models.SET_NULL, null=True, blank=True,
        db_column='entered_by', related_name='harvest_plans_entered',
    )
    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    class Meta:
        db_table = schema_table('export', 'weekly_harvest_plans')
        indexes = [
            models.Index(fields=['status'], name='ix_harvest_plan_status'),
        ]
        constraints = [
            models.UniqueConstraint(
                fields=['season', 'block', 'week_number', 'year'],
                name='uq_weekly_plan',
            ),
            models.CheckConstraint(
                check=(
                    models.Q(monday_plan_kg__gte=0) &
                    models.Q(tuesday_plan_kg__gte=0) &
                    models.Q(wednesday_plan_kg__gte=0) &
                    models.Q(thursday_plan_kg__gte=0) &
                    models.Q(friday_plan_kg__gte=0) &
                    models.Q(saturday_plan_kg__gte=0)
                ),
                name='chk_harvest_plan_kg_gte0',
            ),
        ]
        ordering = ['year', 'week_number', 'block']

    def __str__(self) -> str:
        return f'W{self.week_number}/{self.year} — block {self.block_id} [{self.status}]'


class BlockManagerAssignment(models.Model):
    """Maps a greenhouse_manager user to one or more greenhouse blocks.

    Used by WeeklyHarvestPlanViewSet to enforce per-block write authorization:
    a greenhouse_manager may only create/update harvest plans for blocks they
    are assigned to.

    DDL: export.block_manager_assignments
    UNIQUE (user_id, block_id)
    """

    user = models.ForeignKey(
        'core.User',
        on_delete=models.CASCADE,
        db_column='user_id',
        related_name='block_assignments',
    )
    block = models.ForeignKey(
        'core.GreenhouseBlock',
        on_delete=models.CASCADE,
        db_column='block_id',
        related_name='manager_assignments',
    )
    is_active = models.BooleanField(default=True)
    created_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        db_table = schema_table('export', 'block_manager_assignments')
        constraints = [
            models.UniqueConstraint(
                fields=['user', 'block'],
                name='uq_block_manager_assignment',
            ),
        ]
        ordering = ['user', 'block__code']

    def __str__(self) -> str:
        return f'{self.user.username} → {self.block.code}'


class QuotaAllocation(models.Model):
    """Export quota granted to a firm for a season.

    DDL: export.quota_allocations — UNIQUE (season_id, export_firm_id)
    used_kg is updated by the finance module when shipments are settled.
    """

    season = models.ForeignKey('core.Season', on_delete=models.PROTECT, db_column='season_id')
    export_firm = models.ForeignKey(
        'core.ExportFirm', on_delete=models.PROTECT,
        db_column='export_firm_id', related_name='quota_allocations',
    )
    granted_kg = models.DecimalField(max_digits=12, decimal_places=2)
    used_kg = models.DecimalField(max_digits=12, decimal_places=2, default=0)

    # Warning flags — set True once notification is sent to avoid repeats
    warning_80_sent = models.BooleanField(default=False)
    warning_90_sent = models.BooleanField(default=False)
    warning_95_sent = models.BooleanField(default=False)

    class Meta:
        db_table = schema_table('export', 'quota_allocations')
        constraints = [
            models.UniqueConstraint(
                fields=['season', 'export_firm'],
                name='uq_quota_season_firm',
            ),
            models.CheckConstraint(check=models.Q(used_kg__gte=0), name='chk_quota_used_kg_gte0'),
            models.CheckConstraint(check=models.Q(granted_kg__gt=0), name='chk_quota_granted_kg_gt0'),
        ]
        ordering = ['export_firm__name_en']


class PriceEntry(models.Model):
    """Market tomato price per city per day.

    DDL: export.price_entries — UNIQUE (date, city_id)
    Used by the PricePanel 7-day trend view.
    """

    date = models.DateField()
    city = models.ForeignKey('core.City', on_delete=models.PROTECT, db_column='city_id')
    price_local = models.DecimalField(max_digits=10, decimal_places=2, null=True, blank=True)
    price_usd = models.DecimalField(max_digits=8, decimal_places=4, null=True, blank=True)
    currency = models.CharField(max_length=10, blank=True, default='')
    source = models.CharField(max_length=30, blank=True, default='')
    entered_by = models.ForeignKey(
        'core.User', on_delete=models.SET_NULL, null=True, blank=True,
        db_column='entered_by', related_name='price_entries_entered',
    )
    created_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        db_table = schema_table('export', 'price_entries')
        constraints = [
            models.UniqueConstraint(fields=['date', 'city'], name='uq_price_date_city')
        ]
        ordering = ['-date', 'city']


class DomesticMarketPrice(models.Model):
    """Domestic Turkmenistan market prices from the Satys_bahalar Excel file.

    Tracks daily tomato/vegetable prices at named bazaars (Tolkuçka, Teke bazar, etc.)
    split by price type (bazar/klent/online) and variety. Used to compare
    local TM prices against export destination prices.

    DDL: export.domestic_market_prices
    No unique constraint in DDL — an index covers (date, market_name, price_type, variety_type)
    for fast range queries.
    """

    # === When & where ===
    date = models.DateField()
    market_name = models.CharField(max_length=100, **cyrillic_collation())

    # === Classification ===
    # price_type: 'bazar', 'klent', 'online' — Latin-only codes, no collation needed
    price_type = models.CharField(max_length=30, blank=True, null=True)
    # variety_type: 'tomato_gulpakly', 'tomato_cherry', 'pepper_round', etc.
    variety_type = models.CharField(max_length=30, blank=True, null=True)

    # === Price (TMT per kg) ===
    price = models.DecimalField(max_digits=8, decimal_places=2)

    # === Audit ===
    entered_by = models.ForeignKey(
        'core.User', on_delete=models.SET_NULL, null=True, blank=True,
        db_column='entered_by', related_name='domestic_prices_entered',
    )
    created_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        db_table = schema_table('export', 'domestic_market_prices')
        # Mirror the DDL index for fast date-range queries by market/type/variety
        indexes = [
            models.Index(
                fields=['date', 'market_name', 'price_type', 'variety_type'],
                name='ix_domestic_price',
            )
        ]
        ordering = ['-date', 'market_name']

    def __str__(self) -> str:
        return f'{self.date} | {self.market_name} | {self.variety_type} — {self.price} TMT/kg'

File: frontend/src/mock/planning.ts
import type {
  IWeeklyHarvestPlan,
  IQuotaDashboardItem,
  IPriceEntry,
  IWeeklyTruckAllocation,
  IBlockSummary,
  IDomesticSale,
} from '@/types';

export const MOCK_HARVEST_PLANS: IWeeklyHarvestPlan[] = [
  {
    id: 1, season: 1, season_name: '2025-2026', block: 1, block_code: 'A', block_name: 'A-Ýyladyşhana',
    week_number: 8, year: 2025,
    monday_plan_kg: 3200, tuesday_plan_kg: 3200, wednesday_plan_kg: 3000,
    thursday_plan_kg: 3200, friday_plan_kg: 3100, saturday_plan_kg: 2800,
    monday_actual_kg: 3150, tuesday_actual_kg: 3280, wednesday_actual_kg: 2950,
    thursday_actual_kg: null, friday_actual_kg: null, saturday_actual_kg: null,
    total_plan_kg: 18500, total_actual_kg: 9380,
    status: 'approved', submitted_at: '2025-02-14T08:00:00+05:00', submitted_by_name: 'toyly_b',
    approved_at: '2025-02-14T12:00:00+05:00', approved_by_name: 'Gadam',
    rejected_at: null, rejected_by_name: null, rejection_note: null,
    entered_by_name: 'toyly_b', updated_at: '2025-02-17T08:00:00+05:00',
  },
  {
    id: 2, season: 1, season_name: '2025-2026', block: 2, block_code: 'B', block_name: 'B-Ýyladyşhana',
    week_number: 8, year: 2025,
    monday_plan_kg: 3500, tuesday_plan_kg: 3500, wednesday_plan_kg: 3200,
    thursday_plan_kg: 3500, friday_plan_kg: 3300, saturday_plan_kg: 3000,
    monday_actual_kg: 3600, tuesday_actual_kg: 3450, wednesday_actual_kg: 3250,
    thursday_actual_kg: null, friday_actual_kg: null, saturday_actual_kg: null,
    total_plan_kg: 20000, total_actual_kg: 10300,
    status: 'submitted', submitted_at: '2025-02-15T09:00:00+05:00', submitted_by_name: 'guwanc_k',
    approved_at: null, approved_by_name: null,
    rejected_at: null, rejected_by_name: null, rejection_note: null,
    entered_by_name: 'guwanc_k', updated_at: '2025-02-17T08:00:00+05:00',
  },
  {
    id: 3, season: 1, season_name: '2025-2026', block: 3, block_code: 'C', block_name: 'C-Ýyladyşhana',
    week_number: 8, year: 2025,
    monday_plan_kg: 2800, tuesday_plan_kg: 2800, wednesday_plan_kg: 2600,
    thursday_plan_kg: 2800, friday_plan_kg: 2700, saturday_plan_kg: 2500,
    monday_actual_kg: null, tuesday_actual_kg: null, wednesday_actual_kg: null,
    thursday_actual_kg: null, friday_actual_kg: null, saturday_actual_kg: null,
    total_plan_kg: 16200, total_actual_kg: null,
    status: 'rejected', submitted_at: '2025-02-14T10:00:00+05:00', submitted_by_name: 'geldimyrat_a',
    approved_at: null, approved_by_name: null,
    rejected_at: '2025-02-15T11:00:00+05:00', rejected_by_name: 'Gadam',
    rejection_note: 'Plan values too low for block C capacity',
    entered_by_name: 'geldimyrat_a', updated_at: '2025-02-16T10:00:00+05:00',
  },
];

export const MOCK_QUOTA_DASHBOARD: IQuotaDashboardItem[] = [
  { id: 1, season: 1, season_name: '2025-2026', export_firm: 1, export_firm_name: 'YGT H.J.', granted_kg: 500000, used_kg: 312000, remaining_kg: 188000, used_pct: 62.4, warning_80_sent: false, warning_90_sent: false, warning_95_sent: false },
  { id: 2, season: 1, season_name: '2025-2026', export_firm: 2, export_firm_name: 'Gök Mäkan', granted_kg: 300000, used_kg: 255000, remaining_kg: 45000, used_pct: 85.0, warning_80_sent: true, warning_90_sent: false, warning_95_sent: false },
  { id: 3, season: 1, season_name: '2025-2026', export_firm: 3, export_firm_name: 'Altyn Asyr', granted_kg: 200000, used_kg: 194000, remaining_kg: 6000, used_pct: 97.0, warning_80_sent: true, warning_90_sent: true, warning_95_sent: true },
  { id: 4, season: 1, season_name: '2025-2026', export_firm: 4, export_firm_name: 'Türkmen Eksport', granted_kg: 150000, used_kg: 48000, remaining_kg: 102000, used_pct: 32.0, warning_80_sent: false, warning_90_sent: false, warning_95_sent: false },
  { id: 5, season: 1, season_name: '2025-2026', export_firm: 5, export_firm_name: 'Gündogar', granted_kg: 250000, used_kg: 201000, remaining_kg: 49000, used_pct: 80.4, warning_80_sent: true, warning_90_sent: false, warning_95_sent: false },
];

export const MOCK_TRUCK_ALLOCATIONS: IWeeklyTruckAllocation[] = [
  { id: 1, season: 1, season_name: '2025-2026', week_number: 13, year: 2026, day_of_week: 1, total_planned_kg: 111000, total_trucks_calc: 6.0, russia_trucks: 3, kazakhstan_trucks: 2, gapy_satys_trucks: 1, decided_by_name: 'Gadam', created_at: '2026-03-24T08:00:00+05:00' },
  { id: 2, season: 1, season_name: '2025-2026', week_number: 13, year: 2026, day_of_week: 2, total_planned_kg: 95000, total_trucks_calc: 5.1, russia_trucks: 2, kazakhstan_trucks: 3, gapy_satys_trucks: 0, decided_by_name: 'Gadam', created_at: '2026-03-24T08:00:00+05:00' },
  { id: 3, season: 1, season_name: '2025-2026', week_number: 13, year: 2026, day_of_week: 3, total_planned_kg: 120500, total_trucks_calc: 6.5, russia_trucks: 4, kazakhstan_trucks: 2, gapy_satys_trucks: 1, decided_by_name: 'Gadam', created_at: '2026-03-24T08:00:00+05:00' },
  { id: 4, season: 1, season_name: '2025-2026', week_number: 13, year: 2026, day_of_week: 4, total_planned_kg: 102000, total_trucks_calc: 5.5, russia_trucks: 3, kazakhstan_trucks: 2, gapy_satys_trucks: 0, decided_by_name: null, created_at: '2026-03-24T08:00:00+05:00' },
  { id: 5, season: 1, season_name: '2025-2026', week_number: 13, year: 2026, day_of_week: 5, total_planned_kg: 87000, total_trucks_calc: 4.7, russia_trucks: 2, kazakhstan_trucks: 2, gapy_satys_trucks: 1, decided_by_name: 'Gadam', created_at: '2026-03-24T08:00:00+05:00' },
  { id: 6, season: 1, season_name: '2025-2026', week_number: 13, year: 2026, day_of_week: 6, total_planned_kg: null, total_trucks_calc: null, russia_trucks: 0, kazakhstan_trucks: 0, gapy_satys_trucks: 0, decided_by_name: null, created_at: '2026-03-24T08:00:00+05:00' },
];

export const MOCK_BLOCK_SUMMARY: IBlockSummary[] = [
  { block_id: 1, block_code: 'A', block_name: 'A-Ýyladyşhana', total_plan_kg: 18500, total_actual_kg: 17900, deficit_kg: -600 },
  { block_id: 2, block_code: 'B', block_name: 'B-Ýyladyşhana', total_plan_kg: 20000, total_actual_kg: 20450, deficit_kg: 450 },
  { block_id: 3, block_code: 'C', block_name: 'C-Ýyladyşhana', total_plan_kg: 16200, total_actual_kg: 15800, deficit_kg: -400 },
  { block_id: 4, block_code: 'D', block_name: 'D-Ýyladyşhana', total_plan_kg: 22000, total_actual_kg: 22000, deficit_kg: 0 },
  { block_id: 5, block_code: 'E', block_name: 'E-Ýyladyşhana', total_plan_kg: 19500, total_actual_kg: 18100, deficit_kg: -1400 },
  { block_id: 6, block_code: 'F', block_name: 'F-Ýyladyşhana', total_plan_kg: 17000, total_actual_kg: 17600, deficit_kg: 600 },
  { block_id: 7, block_code: 'G', block_name: 'G-Ýyladyşhana', total_plan_kg: 21000, total_actual_kg: 19500, deficit_kg: -1500 },
  { block_id: 8, block_code: 'H', block_name: 'H-Ýyladyşhana', total_plan_kg: 15000, total_actual_kg: 15200, deficit_kg: 200 },
];

export const MOCK_DOMESTIC_SALES: IDomesticSale[] = [
  { id: 1, date: '2026-03-20', buyer: 1, buyer_name: 'Oraz', block: 1, block_code: 'A', block_name: 'A-Ýyladyşhana', export_firm: null, export_firm_name: null, weight_kg: 450, variety: 'gulpakly', price_per_kg: 3.5, tabel_no: 'T-001', notes: null, created_by_name: 'Gadam', created_at: '2026-03-20T09:00:00+05:00' },
  { id: 2, date: '2026-03-20', buyer: 2, buyer_name: 'Bägül', block: 2, block_code: 'B', block_name: 'B-Ýyladyşhana', export_firm: null, export_firm_name: null, weight_kg: 800, variety: 'cherry', price_per_kg: 4.2, tabel_no: 'T-002', notes: 'Premium grade', created_by_name: 'Gadam', created_at: '2026-03-20T10:30:00+05:00' },
  { id: 3, date: '2026-03-21', buyer: 3, buyer_name: 'Merdan', block: 3, block_code: 'C', block_name: 'C-Ýyladyşhana', export_firm: null, export_firm_name: null, weight_kg: 320, variety: 'gulpakly', price_per_kg: 3.5, tabel_no: 'T-003', notes: null, created_by_name: 'Gadam', created_at: '2026-03-21T08:00:00+05:00' },
  { id: 4, date: '2026-03-22', buyer: 4, buyer_name: 'Aýna', block: 5, block_code: 'E', block_name: 'E-Ýyladyşhana', export_firm: null, export_firm_name: null, weight_kg: 100, variety: null, price_per_kg: null, tabel_no: null, notes: 'Bazar üçin', created_by_name: 'Gadam', created_at: '2026-03-22T11:00:00+05:00' },
  { id: 5, date: '2026-03-23', buyer: 5, buyer_name: 'Döwran', block: 7, block_code: 'G', block_name: 'G-Ýyladyşhana', export_firm: null, export_firm_name: null, weight_kg: 600, variety: 'cherry', price_per_kg: 4.0, tabel_no: 'T-005', notes: null, created_by_name: 'Gadam', created_at: '2026-03-23T09:30:00+05:00' },
];

export const MOCK_PRICE_ENTRIES: IPriceEntry[] = [
  { id: 1, date: '2025-02-17', city: 1, city_name: 'Almaty', price_local: 420, price_usd: 0.95, currency: 'KZT', source: 'market', entered_by_name: 'sales_rep', created_at: '2025-02-17T10:00:00+05:00' },
  { id: 2, date: '2025-02-17', city: 2, city_name: 'Astana', price_local: 410, price_usd: 0.93, currency: 'KZT', source: 'market', entered_by_name: 'sales_rep', created_at: '2025-02-17T10:00:00+05:00' },
  { id: 3, date: '2025-02-17', city: 3, city_name: 'Moscow', price_local: 95, price_usd: 1.05, currency: 'RUB', source: 'market', entered_by_name: 'sales_rep', created_at: '2025-02-17T10:00:00+05:00' },
  { id: 4, date: '2025-02-16', city: 1, city_name: 'Almaty', price_local: 415, price_usd: 0.94, currency: 'KZT', source: 'market', entered_by_name: 'sales_rep', created_at: '2025-02-16T10:00:00+05:00' },
  { id: 5, date: '2025-02-16', city: 2, city_name: 'Astana', price_local: 405, price_usd: 0.92, currency: 'KZT', source: 'market', entered_by_name: 'sales_rep', created_at: '2025-02-16T10:00:00+05:00' },
  { id: 6, date: '2025-02-16', city: 3, city_name: 'Moscow', price_local: 92, price_usd: 1.02, currency: 'RUB', source: 'market', entered_by_name: 'sales_rep', created_at: '2025-02-16T10:00:00+05:00' },
  { id: 7, date: '2025-02-15', city: 1, city_name: 'Almaty', price_local: 408, price_usd: 0.92, currency: 'KZT', source: 'market', entered_by_name: 'sales_rep', created_at: '2025-02-15T10:00:00+05:00' },
];
